src/crop-manager.sol

```diff
@@ -0,0 +1,171 @@
+}
+
+contract UrnProxy {
+    address immutable public vat;
+    address immutable public usr;
+    address public owner;
+
+    constructor(address vat_, address usr_) public {
+        owner = msg.sender;
+        vat = vat_;
+        usr = usr_;
+        VatLike(vat_).hope(msg.sender);
+    }
+
+    function migrate(address newOwner) external {
+        require(msg.sender == owner, "UrnProxy/not-owner");
+        VatLike(vat).hope(newOwner);
+        VatLike(vat).nope(owner);
+        owner = newOwner;
+    }
+}
+
+contract CropJoinManager {
+    mapping (address => uint256) public wards;
+    mapping (address => address) public proxy;  // UrnProxy per user
+    address public implementation;
+
+    event Rely(address indexed);
+    event Deny(address indexed);
+    event SetImplementation(address indexed);
+
+    constructor() public {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    function rely(address usr) public auth { wards[usr] = 1; emit Rely(msg.sender); }
+    function deny(address usr) public auth { wards[usr] = 0; emit Deny(msg.sender); }
+    modifier auth { require(wards[msg.sender] == 1, "CropJoinManager/non-authed"); _; }
+
+
+    function setImplementation(address implementation_) external auth {
+        implementation = implementation_;
+        emit SetImplementation(implementation_);
+    }
+
+    fallback() external {
+        address _impl = implementation;
+        require(_impl != address(0));
+
+        assembly {
+            let ptr := mload(0x40)
+            calldatacopy(ptr, 0, calldatasize())
+            let result := delegatecall(gas(), _impl, ptr, calldatasize(), 0, 0)
+            let size := returndatasize()
+            returndatacopy(ptr, 0, size)
+
+            switch result
+            case 0 { revert(ptr, size) }
+            default { return(ptr, size) }
+        }
+    }
+}
+
+contract CropJoinManagerImp {
+    bytes32 slot0;
+    mapping (address => address) public proxy;  // UrnProxy per user
+
+    address public immutable vat;
+    constructor(address vat_) public {
+        vat = vat_;
+    }
+
+    function getProxy(address usr) internal returns (address urp) {
+        urp = proxy[usr];
+        if (urp == address(0)) {
+            urp = proxy[usr] = address(new UrnProxy(address(vat), usr));
+        }
+    }
+
+    function createProxy() external returns (address) {
+        return getProxy(msg.sender);
+    }
+
+    function join(address crop, address urn, uint256 val) external {
+        TokenLike(CropLike(crop).gem()).transferFrom(msg.sender, address(this), val);
+        TokenLike(CropLike(crop).gem()).approve(crop, val);
+        CropLike(crop).join(getProxy(urn), urn, val);
+    }
+
+    function exit(address crop, address usr, uint256 val) external {
+        CropLike(crop).exit(getProxy(msg.sender), usr, val);
+    }
+
+    function flee(address crop) external {
+        CropLike(crop).flee(getProxy(msg.sender));
+    }
```

Comment:
Does it _really_ make sense to maybe create a proxy in `exit` and `flee`? The only way the calls do anything is if the urn proxy's address has been computed in advance and had `gem` `fluxed` to it and `stake` `tacked` to it. But that address depends on the nonce of the manager contract, which can change between TXs, and I don't think you can read a contract's nonce from within the EVM. So the only way you do it safely would be to send a TX that calls some contract that does the following sequence: 1) `flux`+`tack` to expected proxy address, 2) `exit`/`flee`, 3) require the address of the created proxy to equal the one that was expected. I don't think that's really any better than doing in one TX 1) create proxy, 2) `flux`+`tack`, 3) `exit`/`flee`. Am I missing something here?

Reply:
Conclusion: just a lookup and a `require`.

Reply:
Done.